Solutions/Marain.UserNotifications.Client/Menes/Hal/WebLink.cs

```diff
@@ -60,6 +60,10 @@ namespace Menes.Hal
         public WebLink(in JsonElement source)
         {
             this.source = source;
+            if (string.IsNullOrEmpty(this.Href))
+            {
+                throw new ArgumentException(nameof(this.Href));
+            }
```

Comment:
This isn't right for three reasons.

First, if the `href` property is entirely absent, this fails with a `KeyNotFoundException`. (That's what `this.source.GetProperty(HrefProperty)` will end up throwing if there is no `href` property.)

Second, this will cause the allocation of an extra `string` on the GC heap. Each time you call `GetString` on a `JsonElement` it will create a new copy of the string and return that.

Third, the single-argument form of `ArgumentException` expects the string to be a message describing the problem, but you've just passed the text `"Href"` here (which is what `nameof(this.Href)` evaluates to). You should pass an error message, and also a second argument identifying the name of the argument that is wrong. And that should be the name of an actual argument to the constructor, and not the name of some property in the type. (Also, if you are going to mention the missing property in the message, which is a a good thing to do, the actual serialized form is `href`, not `Href`.)


Try this:

```cs
public WebLink(in JsonElement source)
{
    if (!source.TryGetProperty(HrefProperty, out JsonElement hrefProperty))
    {
        throw new ArgumentException("WebLink must have an href", nameof(source));
    }
    if (hrefProperty.ValueKind != JsonValueKind.String || hrefProperty.ValueEquals(string.Empty))
    {
        throw new ArgumentException("WebLink href must be a non-empty string", nameof(source));
    }

    this.source = source;
}
```

We use `TryGetProperty` to detect the case where the property is entirely absent and throw an exception with a suitable message for that scenario. And then when the property is present, we first check that it's a string, in case someone has inexplicably tried to write something like `{"href": false}`. Note that checking for `ValueKind` being `JsonValueKind.String` will also weed out `{"href":null}` because that would show up as `JsonValueKind.Null`.

And then finally, we don't actually need to inspect the string, we just need to know that it's non-empty. And the most efficient way to do that is to compare it with `string.Empty` because it then doesn't need to materialize an actual `string` instance containing the value. (The way `JsonElement` works is that if at all possible, it just maintains pointers into the original JSON document in UTF8 form. So the `JsonElement` we get back from `TryGetProperty` is really just a numbers indicating the offset into that UTF8 text. e.g.:

![image](https://user-images.githubusercontent.com/334181/155968099-e7683cef-75c0-464a-87ec-7bd7def0bfad.png)

It tries to avoid allocating any additional memory. But as soon as you call `GetString` you've forced it into returning you a `string` containing the property value. So it has to allocate space for that on the GC heap.

But if you call `ValueEquals` it can compare the property value in situ in the UTF8 document with the string passed in. In this case I've passed `string.Empty`, which is a singleton, so that won't ever allocate eitther.

(I was hoping there'd be a simpler way, e.g., asking the property how long it is, but apparently not.)